pandas/core/reshape/merge.py

```diff
@@ -1604,7 +1604,10 @@ def _factorize_keys(lk, rk, sort=True):
 
         lk = ensure_int64(lk.codes)
         rk = ensure_int64(rk)
-    elif is_int_or_datetime_dtype(lk) and is_int_or_datetime_dtype(rk):
+    elif (issubclass(lk.dtype.type, (np.integer, np.timedelta64,
```

Comment:
this is not right, these should be in separate branches here, the datetime / timedeltas can be in a single elif clause.

Reply:
If the behavior here is wrong then we should also add a test for it.  I'll open a separate issue to address this, since I'm not sure off the top of my head what the correct behavior is.

Reply:
opened #23929

Reply:
OK.  I'd really rather these be separated in a dedicated PR, but I guess #23929 can be a Needs Test Issue... will change.

Reply:
yeah that's fine. I think there is another PR about adding EA types here as well (though of course orthogonal). see if this passes and can add tests / perf checking later.